Support no_std where possible (specifically for wasm32v1-none)

**Is your feature request related to a problem? Please describe.**
`naga`, `wgpu` and related crates should have `std` as a (default) feature, specifically for `wasm32v1-none` support.

**Describe the solution you'd like**
There's a hope to get the wasm ecosystem slowly shifting from `wasm32-unknown-unknown` to [`wasm32v1-none`](https://doc.rust-lang.org/nightly/rustc/platform-support/wasm32v1-none.html), which is hopefully stabilizing in rust 1.84. This is primarily because `wasm32-unknown-unknown` is [something of a moving target and can change the browser requirements at will](https://blog.rust-lang.org/2024/09/24/webassembly-targets-change-in-default-target-features.html), which we recently discovered when older Safari stopped working 😅

The main differences between `wasm32v1-none` and `wasm32-unknown-unknown` are:
- It does not enable any wasm features by default (which means it won't arbitrarily break browser compatibility with rust updates)
- It has no std (as opposed to a stubbed pseudo-std)

Unfortunately, no_std is both a blessing and a curse: it's neat to get things using core/alloc where possible, but it sucks that the majority of the ecosystem needs churn to make this viable.

I've done a quick investigation and I think it's plausible to get at least web working in `no_std`, but potentially other platforms too. `wgpu-types` is the easiest to get `no_std`, it only needs it for one specific dx12 config. `naga` is possibly the second easiest crate, but it requires some work in dependencies. wgpu-core and wgpu-hal will need more work mostly surrounding the use of mutexes.

**Describe alternatives you've considered**
We could just _not_, but that means it's impossible to allow users of wgpu to have any control over which browsers they support.
We can also limit this to a specific set of features, there's no real need today for DX12 and `no_std`, for example.

There is definitely work remaining to be done in order to meet the goal of running wgpu, with a web backend, on `wasm32v1-none`. It has many dependencies that are not `no_std`. For example, here is are some of the errors from a simple `cargo check`:

```
$ cargo check -p wgpu --no-default-features --target=wasm32v1-none
error[E0463]: can't find crate for `std`
  --> /Users/kpreid/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/raw-window-handle-0.6.2/src/lib.rs:34:1
   |
34 | extern crate std;
   | ^^^^^^^^^^^^^^^^^ can't find crate
   |
   = note: the `wasm32v1-none` target may not support the standard library

error[E0463]: can't find crate for `std`
  |
  = note: the `wasm32v1-none` target may not support the standard library
  = note: `std` is required by `parking_lot_core` because it does not declare `#![no_std]`

error[E0463]: can't find crate for `std`
   --> /Users/kpreid/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/thiserror-2.0.11/src/lib.rs:275:1
    |
275 | extern crate std;
    | ^^^^^^^^^^^^^^^^^ can't find crate
    |
    = note: the `wasm32v1-none` target may not support the standard library
```

(Note that because these errors are in different crates, any single build command may abort early and not print the rest of them.)

This means that you cannot even use the WebGPU backend. If we also want to support WebGL, then additional work is needed:

* `naga` needs `no_std` support, including removing or PRing std dependencies like `termcolor`
* The `gles` backend needs to be modified to, when working with `webgl`, avoid using `Mutex` and use `RefCell` instead, and other such changes.

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -327,6 +327,7 @@ jobs:
           cargo --locked clippy --target ${{ matrix.target }} ${{ matrix.extra-flags }} -p wgpu-sync --no-default-features
           cargo --locked clippy --target ${{ matrix.target }} ${{ matrix.extra-flags }} -p naga --no-default-features
           cargo --locked clippy --target ${{ matrix.target }} ${{ matrix.extra-flags }} -p wgpu-hal --no-default-features
+          cargo --locked clippy --target ${{ matrix.target }} ${{ matrix.extra-flags }} -p wgpu-core --no-default-features
           cargo --locked clippy --target ${{ matrix.target }} ${{ matrix.extra-flags }} -p wgpu --no-default-features
 
           # Check with all compatible features
@@ -335,6 +336,7 @@ jobs:
           cargo --locked clippy --target ${{ matrix.target }} ${{ matrix.extra-flags }} -p wgpu-sync --no-default-features --features alloc
           cargo --locked clippy --target ${{ matrix.target }} ${{ matrix.extra-flags }} -p naga --no-default-features --features dot-out,spv-in,spv-out,glsl-out,msl-out,serialize,deserialize,wgsl-in,wgsl-out,hlsl-out
           cargo --locked clippy --target ${{ matrix.target }} ${{ matrix.extra-flags }} -p wgpu-hal --no-default-features --features fragile-send-sync-non-atomic-wasm
+          cargo --locked clippy --target ${{ matrix.target }} ${{ matrix.extra-flags }} -p wgpu-core --no-default-features --features api_log_info,resource_log_info,strict_asserts,serde,wgsl,spirv,counters,noop
           cargo --locked clippy --target ${{ matrix.target }} ${{ matrix.extra-flags }} -p wgpu --no-default-features --features serde
 
       # Building for native platforms with standard tests.

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -337,7 +337,7 @@ jobs:
           cargo --locked clippy --target ${{ matrix.target }} ${{ matrix.extra-flags }} -p naga --no-default-features --features dot-out,spv-in,spv-out,glsl-out,msl-out,serialize,deserialize,wgsl-in,wgsl-out,hlsl-out
           cargo --locked clippy --target ${{ matrix.target }} ${{ matrix.extra-flags }} -p wgpu-hal --no-default-features --features fragile-send-sync-non-atomic-wasm
           cargo --locked clippy --target ${{ matrix.target }} ${{ matrix.extra-flags }} -p wgpu-core --no-default-features --features api_log_info,resource_log_info,strict_asserts,serde,wgsl,spirv,counters,noop
-          cargo --locked clippy --target ${{ matrix.target }} ${{ matrix.extra-flags }} -p wgpu --no-default-features --features serde
+          cargo --locked clippy --target ${{ matrix.target }} ${{ matrix.extra-flags }} -p wgpu --no-default-features --features webgpu,drm,wgsl,strict_asserts,serde,counters
 
       # Building for native platforms with standard tests.
       - name: Check native

--- a/wgpu-core/Cargo.toml
+++ b/wgpu-core/Cargo.toml
@@ -111,7 +111,7 @@ fragile-send-sync-non-atomic-wasm = [
 
 ## Enable certain items to be `Send` and `Sync` when they would not otherwise be.
 ## Also enables backtraces in some error cases when also under cfg(debug_assertions).
-std = ["wgpu-sync/std"]
+std = ["wgpu-sync/std", "once_cell/std"]
 
 #! ### External libraries
 # --------------------------------------------------------------------
@@ -199,12 +199,13 @@ hashbrown.workspace = true
 indexmap.workspace = true
 log.workspace = true
 macro_rules_attribute = { workspace = true, optional = true }
-once_cell = { workspace = true, features = ["std"] }
+num-traits = { workspace = true }
+once_cell = { workspace = true, default-features = false }
 profiling = { workspace = true, default-features = false }
 raw-window-handle.workspace = true
 ron = { workspace = true, optional = true }
 rustc-hash.workspace = true
-serde = { workspace = true, features = ["default", "derive"], optional = true }
+serde = { workspace = true, features = ["derive"], default-features = false, optional = true }
 smallvec.workspace = true
 thiserror.workspace = true
 

--- a/wgpu-core/src/pool.rs
+++ b/wgpu-core/src/pool.rs
@@ -2,10 +2,14 @@ use alloc::sync::{Arc, Weak};
 use core::hash::Hash;
 
 use hashbrown::{hash_map::Entry, HashMap};
-use once_cell::sync::OnceCell;
 
 use crate::lock::{rank, Mutex};
 use crate::FastHashMap;
+
+#[cfg(feature = "std")]
+use once_cell::sync::OnceCell;
+#[cfg(not(feature = "std"))]
+use once_cell::unsync::OnceCell;
 
 type SlotInner<V> = Weak<V>;
 type ResourcePoolSlot<V> = Arc<OnceCell<SlotInner<V>>>;

--- a/wgpu-core/src/timestamp_normalization/mod.rs
+++ b/wgpu-core/src/timestamp_normalization/mod.rs
@@ -391,6 +391,9 @@ impl TimestampNormalizationBindGroup {
 }
 
 fn compute_timestamp_period(input: f32) -> (u32, u32) {
+    #[cfg(not(feature = "std"))]
+    use num_traits::float::Float as _;
+
     let pow2 = input.log2().ceil() as i32;
     let clamped_pow2 = pow2.clamp(-32, 32).unsigned_abs();
     let shift = 32 - clamped_pow2;

